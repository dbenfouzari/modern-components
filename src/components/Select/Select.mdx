---
name: Select
---

import { Playground, PropsTable } from 'docz';
import Select from './';

# Select

<PropsTable of={Select} />

## Basic usage

<Playground>
  {() => {
    class Example extends React.Component {
      constructor (props) {
        super(props);
        this.state = {
          value: []
        }
        this.handleChange = this.handleChange.bind(this);
      }

      handleChange(nextValue) {
        this.setState({ value: nextValue });
      }
      
      render () {
        return (
          <Select
            options={
              [
                { value: 1, label: "One" },
                { value: 2, label: "Two" },
                { value: 3, label: "Three" }
              ]
            }
            value={this.state.value}
            onChange={this.handleChange}
          />
        )
      }
    }
    
    return <Example />;
  }}
</Playground>
