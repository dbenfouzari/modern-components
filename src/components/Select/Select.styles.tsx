import { lighten } from "polished";
import * as React from "react";
import styled, { css } from "../../styled-components";

export const SelectWrapper = styled.div`
  position: relative;
`;

export const InputWrapper = styled(({ isOpen, ...rest }: { isOpen: boolean }) => <div {...rest} />)`
  border-radius: 4px;
  border: 1px solid ${(props) => props.theme.primaryColor};
  display: flex;
  padding: 5px;

  ${(props) => props.isOpen ? css`
    border-bottom: none;
    border-radius: 4px 4px 0 0;
  ` : null}
`;

export const Tag = styled.span`
  background-color: ${(props) => lighten(0.3, props.theme.primaryColor)};
  border-radius: 2px;
  border: 1px solid ${(props) => props.theme.primaryColor};
  font-size: 0.8em;
  padding: 2px 15px 2px 2px;
  position: relative;

  & > span {
    cursor: default;
    position: absolute;
    right: 4px;
  }

  & + & {
    margin-left: 5px;
  }
`;

export const Input = styled.input`
  flex: 1;
  border: 0;
  outline: none;
`;

export const OptionsWrapper = styled.ul`
  background-color: #fff;
  border-radius: 0 0 4px 4px;
  border: 1px solid ${(props) => props.theme.primaryColor};
  border-top: none;
  box-sizing: border-box;
  list-style: none;
  margin: 0;
  padding: 0;
  position: absolute;
  top: calc(100% - 1px);
  width: 100%;
`;

export const Option = styled.li<{isFocused?: boolean}>`
  cursor: default;
  padding: 10px;

  ${(props) => props.isFocused ? css`
    background-color: ${lighten(0.2, props.theme.primaryColor)};
  ` : null}

  ${(props) => props.onMouseDown ? css`
    &:hover {
      background-color: ${lighten(0.2, props.theme.primaryColor)};
    }
  ` : null}
`;
