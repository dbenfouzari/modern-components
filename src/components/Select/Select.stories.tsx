import { action } from "@storybook/addon-actions";
import { storiesOf } from "@storybook/react";
import * as React from "react";

import Select from ".";
import { OptionType } from "./Select";

const options: OptionType[] = [
  { value: 1, label: "One" },
  { value: 2, label: "Two" },
  { value: 3, label: "Three" },
];

storiesOf("Select", module)
  .add("with text", () => {
    const handleChange = (nextValue: Array<OptionType["value"]>) => {
      action("selectChange")(nextValue);
    };

    return (
        <Select options={options} value={[1]} onChange={handleChange} />
    );
  });
