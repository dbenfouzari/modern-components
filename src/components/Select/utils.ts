import { OptionType } from "./Select";

export const mapValueToLabelForOptions = (options: OptionType[]) => (value: OptionType["value"]) => {
  const opt = options.find((option) => option.value === value);

  if (opt) { return opt.label; }
  return "???";
};
