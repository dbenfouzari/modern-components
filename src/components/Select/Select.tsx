import * as React from "react";
import {
  Input,
  InputWrapper,
  Option,
  OptionsWrapper,
  SelectWrapper,
  Tag
} from "./Select.styles";
import { mapValueToLabelForOptions } from "./utils";

interface OptionType {
  value: string | number;
  label: string | number;
}

interface KeyCodes {
  [asString: string]: number;
}

const KEY_CODES: KeyCodes = {
  BACKSPACE: 8,
  BOTTOM: 40,
  ENTER: 13,
  UP: 38
};

interface SelectProps {
  /**
   * The value
   *
   * @since 0.0.1
   * @version 0.0.1
   */
  value: Array<OptionType["value"]>;
  /**
   * The handler
   *
   * @since 0.0.1
   * @version 0.0.1
   */
  onChange: (nextValue: Array<OptionType["value"]>) => void;
  /**
   * This is the list of options you want to show to the user
   *
   * @since 0.0.1
   * @version 0.0.1
   */
  options: OptionType[];
}

interface SelectState {
  currentInput: string;
  focusedOptionIndex: number;
  isOpen: boolean;
}

class Select extends React.Component<SelectProps, SelectState> {
  public state = {
    currentInput: "",
    focusedOptionIndex: -1,
    isOpen: false
  };

  public get notSelectedOptions(): OptionType[] {
    return this.props.options.filter(
      option => !this.props.value.includes(option.value)
    );
  }

  public get valueFilteredOptions(): OptionType[] {
    return this.notSelectedOptions.filter(option => {
      const regex = new RegExp(this.state.currentInput, "gi");
      return regex.test(option.label.toString());
    });
  }

  public onOptionRemove = (optionId: OptionType["value"]) => () => {
    const nextValue = [...this.props.value];
    const optionIndex = this.props.value.findIndex(val => val === optionId);

    nextValue.splice(optionIndex, 1);

    this.props.onChange(nextValue as any);
  };

  public handleFocus = () => {
    this.setState({ isOpen: true });
  };

  public handleBlur = () => {
    this.setState({ isOpen: false });
  };

  public handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    switch (event.keyCode) {
      case KEY_CODES.BACKSPACE:
        if (event.currentTarget.value !== "") {
          return;
        }

        // Backspace pressed, but no value in the input.
        // So we should remove tag
        if (this.props.value.length) {
          const nextValue = [...this.props.value];
          nextValue.pop();
          this.props.onChange(nextValue);
        }
        break;

      case KEY_CODES.BOTTOM:
        event.preventDefault();
        if (this.valueFilteredOptions.length) {
          const targetIndex = this.state.focusedOptionIndex + 1;
          let target;

          if (this.valueFilteredOptions.length - 1 >= targetIndex) {
            target = targetIndex;
          } else {
            target = 0;
          }

          this.setState({ focusedOptionIndex: target });
        }
        break;

      case KEY_CODES.UP:
        event.preventDefault();
        if (this.valueFilteredOptions.length) {
          const targetIndex = this.state.focusedOptionIndex - 1;
          let target;

          if (targetIndex < 0) {
            target = this.valueFilteredOptions.length - 1;
          } else if (this.valueFilteredOptions.length - 1 >= targetIndex) {
            target = targetIndex;
          } else {
            target = targetIndex;
          }

          this.setState({ focusedOptionIndex: target });
        }
        break;

      case KEY_CODES.ENTER:
        event.preventDefault();
        if (this.state.focusedOptionIndex > -1) {
          const focusedOptionValue = this.valueFilteredOptions[
            this.state.focusedOptionIndex
          ].value;
          const nextValue = [...this.props.value, focusedOptionValue];
          this.props.onChange(nextValue);

          this.setState({
            currentInput: "",
            focusedOptionIndex: -1
          });
        }
        break;

      default:
        break;
    }
  };

  public mapLabelToSearch = (label: OptionType["label"]) => {
    if (this.state.currentInput) {
      const regex = new RegExp(`(${this.state.currentInput})`, "gi");
      return label.toString().replace(regex, "<b>$1</b>");
    }

    return label.toString();
  };

  public handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ currentInput: event.currentTarget.value });
  };

  public handleOptionSelect = (optionId: OptionType["value"]) => () => {
    const nextValue = [...this.props.value];
    nextValue.push(optionId);
    this.props.onChange(nextValue);
  };

  public render() {
    const { isOpen, currentInput, focusedOptionIndex } = this.state;

    const mapValueToLabel = mapValueToLabelForOptions(this.props.options);

    return (
      <SelectWrapper>
        <InputWrapper isOpen={isOpen}>
          {this.props.value.map(value => (
            <Tag key={value}>
              {mapValueToLabel(value)}
              <span onClick={this.onOptionRemove(value)}>&times;</span>
            </Tag>
          ))}

          <Input
            value={currentInput}
            onChange={this.handleInputChange}
            onFocus={this.handleFocus}
            onBlur={this.handleBlur}
            onKeyDown={this.handleKeyDown}
          />
        </InputWrapper>

        {isOpen ? (
          <OptionsWrapper>
            {this.valueFilteredOptions.length > 0 ? (
              this.valueFilteredOptions.map((option, optionIndex) => (
                <Option
                  key={option.value}
                  onMouseDown={this.handleOptionSelect(option.value)}
                  isFocused={focusedOptionIndex === optionIndex}
                  dangerouslySetInnerHTML={{
                    __html: this.mapLabelToSearch(option.label)
                  }}
                />
              ))
            ) : (
              <Option>No result</Option>
            )}
          </OptionsWrapper>
        ) : null}
      </SelectWrapper>
    );
  }
}

export default Select;
