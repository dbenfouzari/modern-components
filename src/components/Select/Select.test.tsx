import * as React from "react";
import renderer from "react-test-renderer";
import { ThemeProvider } from "../../styled-components";
import defaultTheme from "../../theme";

import Select from ".";

describe("Select", () => {
  it("should render successfully", () => {
    const handleChange = jest.fn();
    const wrapper = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <Select
            options={[{ value: 1, label: "One" }]}
            onChange={handleChange}
            value={[1]}
          />
        </ThemeProvider>,
      )
      .toJSON();

    expect(wrapper).toMatchSnapshot();
  });
});
