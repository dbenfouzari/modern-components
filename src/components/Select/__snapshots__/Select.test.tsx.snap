// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render successfully 1`] = `
.c0 {
  position: relative;
}

.c1 {
  border-radius: 4px;
  border: 1px solid #3498DB;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 5px;
}

.c5 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  border: 0;
  outline: none;
}

.c3 {
  background-color: #3498DB;
  border-radius: 2px;
  border: 1px solid #3498DB;
  font-size: 0.8em;
  padding: 2px 15px 2px 2px;
  position: relative;
}

.c3 > span {
  cursor: default;
  position: absolute;
  right: 4px;
}

.c3 + .c4 {
  margin-left: 5px;
}

.c3 + .c2 {
  margin-left: 5px;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <span
      className="c2 c3"
    >
      One
      <span
        onClick={[Function]}
      >
        Ã—
      </span>
    </span>
    <input
      className="c4 c5"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      value=""
    />
  </div>
</div>
`;
